File: converter.js
const fs = require('fs');
const path = require('path');
const Anthropic = require('@anthropic-ai/sdk');
const { app } = require('electron');

// Get paths
const SKILLS_PATH = path.join(__dirname, 'skills');
const PROMPTS_PATH = path.join(__dirname, 'prompts');

// Model pricing (per million tokens)
const MODEL_PRICING = {
  'claude-haiku-4-5': { input: 1.00, output: 5.00 },
  'claude-sonnet-4-5-20250929': { input: 3.00, output: 15.00 },
  'claude-3-5-haiku-20241022': { input: 0.80, output: 4.00 }
};

/**
 * Load prompt based on mode (advanced, simple, or custom)
 */
function loadMasterPrompt(settings, outputPath, fileName) {
  let basePrompt;

  // Load appropriate prompt file based on mode
  if (settings.promptMode === 'custom') {
    // Custom prompt should be passed in settings.customPrompt
    basePrompt = settings.customPrompt || fs.readFileSync(
      path.join(PROMPTS_PATH, 'basic-prompt.txt'),
      'utf-8'
    );
  } else if (settings.promptMode === 'advanced') {
    basePrompt = fs.readFileSync(
      path.join(PROMPTS_PATH, 'master-prompt.txt'),
      'utf-8'
    );
  } else {
    // Default to simple
    basePrompt = fs.readFileSync(
      path.join(PROMPTS_PATH, 'basic-prompt.txt'),
      'utf-8'
    );
  }

  // Build skills documentation reference (only for advanced mode)
  let skillsSection = '';
  if (settings.promptMode === 'advanced') {
    skillsSection = `

**Skills Documentation Available:**
You have access to the following documentation for reference:
- ${path.join(SKILLS_PATH, 'docx/SKILL.md')}
- ${path.join(SKILLS_PATH, 'docx/docx-js.md')}
- ${path.join(SKILLS_PATH, 'pdf/SKILL.md')}

Read these files before generating code if needed for clarification.
`;
  }

  // Build enhanced prompt with settings
  const enhancedPrompt = `
${basePrompt}

---

## APPLICATION SETTINGS

**Output Configuration:**
- Save file to: ${outputPath}/${fileName}.docx
- Font: ${settings.font}
- Font Size: ${settings.fontSize}pt (${settings.fontSize * 2} in half-points)
- Margins: Top ${settings.margins.top / 1440}", Right ${settings.margins.right / 1440}", Bottom ${settings.margins.bottom / 1440}", Left ${settings.margins.left / 1440}"

**Special Requests:**
${settings.specialRequests.replaceSignatures ? '- ✅ Replace signatures with [Signature]' : '- ❌ Keep signature images as-is'}
${settings.specialRequests.addPageMarkers ? '- ✅ Add [Page X of the original] markers (except page 1)' : '- ❌ Do not add page markers'}
${skillsSection}
---

## CODE GENERATION INSTRUCTIONS

**CRITICAL: Return ONLY executable JavaScript code.**

1. Generate complete, runnable Node.js code using docx.js
2. Include ALL necessary require() statements
3. Include file I/O (fs.writeFileSync)
4. Wrap code in \`\`\`javascript code blocks
5. Print "SUCCESS: ${fileName}.docx" when complete
6. Handle errors with try/catch
7. Exit with process.exit(0) on success, process.exit(1) on error

**Example code structure:**
\`\`\`javascript
const fs = require('fs');
const { Document, Packer, Paragraph, TextRun, Table, TableRow, TableCell,
        AlignmentType, BorderStyle, WidthType, ShadingType, PageBreak } = require('docx');

try {
  // Your document creation code here
  const doc = new Document({
    creator: "",
    description: "",
    title: "",
    // ... document structure
  });
  
  Packer.toBuffer(doc).then(buffer => {
    fs.writeFileSync('${outputPath}/${fileName}.docx', buffer);
    console.log('SUCCESS: ${fileName}.docx');
    process.exit(0);
  }).catch(err => {
    console.error('ERROR:', err.message);
    process.exit(1);
  });
} catch (error) {
  console.error('ERROR:', error.message);
  process.exit(1);
}
\`\`\`

**Now generate the code for this document.**
`;

  return enhancedPrompt;
}

/**
 * Convert file to base64 for API
 */
function fileToBase64(filePath) {
  const fileBuffer = fs.readFileSync(filePath);
  return fileBuffer.toString('base64');
}

/**
 * Determine media type from file extension
 */
function getMediaType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const types = {
    '.pdf': 'application/pdf',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.png': 'image/png'
  };
  return types[ext] || 'application/pdf';
}

/**
 * Extract JavaScript code from Claude's response
 */
function extractCode(responseText) {
  // Find code blocks
  const codeBlockRegex = /```javascript\n([\s\S]*?)\n```/g;
  const matches = [...responseText.matchAll(codeBlockRegex)];
  
  if (matches.length === 0) {
    // Try without language specifier
    const genericCodeRegex = /```\n([\s\S]*?)\n```/g;
    const genericMatches = [...responseText.matchAll(genericCodeRegex)];
    
    if (genericMatches.length > 0) {
      return genericMatches.map(m => m[1]).join('\n\n');
    }
    
    throw new Error('No code blocks found in response');
  }
  
  // Combine all code blocks
  return matches.map(m => m[1]).join('\n\n');
}

/**
 * Execute generated code
 */
async function executeCode(code, tempDir, fileName, progressCallback) {
  const codePath = path.join(tempDir, `${fileName}_converter.js`);

  // Write code to temp file
  fs.writeFileSync(codePath, code);

  console.log(`[DEBUG] Executing code from: ${codePath}`);
  console.log(`[DEBUG] Generated code:\n${code.substring(0, 500)}...`);

  return new Promise((resolve, reject) => {
    const { spawn } = require('child_process');

    // Set NODE_PATH to include the project's node_modules
    const projectNodeModules = path.join(__dirname, 'node_modules');
    const env = {
      ...process.env,
      NODE_PATH: projectNodeModules
    };

    const nodeProcess = spawn('node', [codePath], { env });

    let stdout = '';
    let stderr = '';

    nodeProcess.stdout.on('data', (data) => {
      const output = data.toString();
      stdout += output;
      console.log(`[STDOUT] ${output}`);

      // Check for success message
      if (output.includes('SUCCESS:')) {
        progressCallback({
          status: 'complete',
          progress: 100
        });
      }
    });

    nodeProcess.stderr.on('data', (data) => {
      const errOutput = data.toString();
      stderr += errOutput;
      console.error(`[STDERR] ${errOutput}`);
    });

    nodeProcess.on('close', (exitCode) => {
      console.log(`[DEBUG] Process exited with code: ${exitCode}`);
      console.log(`[DEBUG] Full stdout: ${stdout}`);
      console.log(`[DEBUG] Full stderr: ${stderr}`);

      // Don't clean up temp file immediately for debugging
      // try {
      //   fs.unlinkSync(codePath);
      // } catch (e) {
      //   // Ignore cleanup errors
      // }

      if (exitCode === 0) {
        resolve({ success: true, output: stdout });
      } else {
        reject(new Error(`Code execution failed (exit code ${exitCode}):\nSTDERR: ${stderr}\nSTDOUT: ${stdout}`));
      }
    });
  });
}

/**
 * Calculate cost from API response usage
 */
function calculateCost(usage, model) {
  const pricing = MODEL_PRICING[model];
  if (!pricing || !usage) return 0;

  const inputCost = (usage.input_tokens / 1_000_000) * pricing.input;
  const outputCost = (usage.output_tokens / 1_000_000) * pricing.output;

  return inputCost + outputCost;
}

/**
 * Parse page range string into array of page numbers
 * Example: "1-5, 7, 9-12" -> [1, 2, 3, 4, 5, 7, 9, 10, 11, 12]
 */
function parsePageRange(rangeStr, totalPages) {
  const pages = new Set();
  const parts = rangeStr.split(',').map(s => s.trim()).filter(s => s.length > 0);

  for (const part of parts) {
    if (part.includes('-')) {
      // Range like "1-5"
      const [startStr, endStr] = part.split('-').map(s => s.trim());
      const start = parseInt(startStr);
      const end = parseInt(endStr);

      if (isNaN(start) || isNaN(end)) {
        throw new Error(`Invalid page range: "${part}". Pages must be numbers.`);
      }
      if (start < 1 || end > totalPages) {
        throw new Error(`Page range "${part}" is out of bounds. Document has ${totalPages} pages.`);
      }
      if (start > end) {
        throw new Error(`Invalid page range: "${part}". Start page must be less than or equal to end page.`);
      }

      for (let i = start; i <= end; i++) {
        pages.add(i);
      }
    } else {
      // Single page like "7"
      const page = parseInt(part);
      if (isNaN(page)) {
        throw new Error(`Invalid page number: "${part}". Must be a number.`);
      }
      if (page < 1 || page > totalPages) {
        throw new Error(`Page ${page} is out of bounds. Document has ${totalPages} pages.`);
      }
      pages.add(page);
    }
  }

  if (pages.size === 0) {
    throw new Error('No valid pages specified in range.');
  }

  return Array.from(pages).sort((a, b) => a - b);
}

/**
 * Extract specific pages from a PDF and create a new PDF file
 * Returns path to the temporary PDF file with selected pages
 */
async function extractPdfPages(inputPath, pageNumbers) {
  const { PDFDocument } = require('pdf-lib');

  console.log(`[PAGE EXTRACTION] Extracting pages ${pageNumbers.join(', ')} from PDF`);

  // Load the PDF
  const inputPdfBytes = fs.readFileSync(inputPath);
  const inputPdf = await PDFDocument.load(inputPdfBytes);

  // Create new PDF with selected pages
  const outputPdf = await PDFDocument.create();

  // Copy selected pages (pdf-lib uses 0-based indexing, so subtract 1)
  for (const pageNum of pageNumbers) {
    const [copiedPage] = await outputPdf.copyPages(inputPdf, [pageNum - 1]);
    outputPdf.addPage(copiedPage);
  }

  // Save to temp file
  const outputBytes = await outputPdf.save();
  const tempPath = path.join(
    app.getPath('temp'),
    `pdf-converter-selected-${Date.now()}.pdf`
  );
  fs.writeFileSync(tempPath, outputBytes);

  console.log(`[PAGE EXTRACTION] Created temporary PDF at: ${tempPath}`);
  return tempPath;
}

/**
 * Main conversion function
 */
async function convertFile(filePath, fileName, settings, apiKey, progressCallback) {
  const client = new Anthropic({ apiKey });
  const tempDir = path.join(app.getPath('temp'), 'pdf-converter');

  // Use the same directory as the input file
  const outputDir = path.dirname(filePath);

  // Ensure temp directory exists
  if (!fs.existsSync(tempDir)) {
    fs.mkdirSync(tempDir, { recursive: true });
  }

  let totalCost = 0;
  let processedFilePath = filePath;
  let tempFileCleanup = null;

  try {
    // Step 1: Prepare file
    progressCallback({ status: 'preparing', progress: 10 });

    const mediaType = getMediaType(filePath);

    // Handle page selection for PDFs
    if (mediaType === 'application/pdf' && settings.pageSelection && settings.pageSelection.mode === 'range') {
      try {
        const rangeStr = settings.pageSelection.range?.trim();

        if (!rangeStr) {
          throw new Error('Page range is empty. Please enter a valid range (e.g., "1-5, 7, 9-12").');
        }

        // Get total page count using pdf-parse
        const pdfParse = require('pdf-parse');
        const dataBuffer = fs.readFileSync(filePath);
        const pdfData = await pdfParse(dataBuffer);
        const totalPages = pdfData.numpages;

        console.log(`[PAGE SELECTION] PDF has ${totalPages} total pages`);

        // Parse and validate range
        const selectedPages = parsePageRange(rangeStr, totalPages);

        console.log(`[PAGE SELECTION] User selected ${selectedPages.length} pages: ${selectedPages.join(', ')}`);

        // Extract selected pages to temporary PDF
        processedFilePath = await extractPdfPages(filePath, selectedPages);

        // Set cleanup function to delete temp file later
        tempFileCleanup = () => {
          try {
            if (fs.existsSync(processedFilePath)) {
              fs.unlinkSync(processedFilePath);
              console.log(`[CLEANUP] Deleted temporary PDF: ${processedFilePath}`);
            }
          } catch (e) {
            console.warn(`[CLEANUP] Could not delete temp file: ${e.message}`);
          }
        };

      } catch (error) {
        throw new Error(`Page selection failed: ${error.message}`);
      }
    }

    const fileBase64 = fileToBase64(processedFilePath);
    const masterPrompt = loadMasterPrompt(settings, outputDir, fileName);

    // Step 2: Send to Claude
    progressCallback({ status: 'analyzing', progress: 30 });

    const response = await client.messages.create({
      model: settings.model,
      max_tokens: 16000,
      messages: [{
        role: 'user',
        content: [
          {
            type: 'document',
            source: {
              type: 'base64',
              media_type: mediaType,
              data: fileBase64
            }
          },
          {
            type: 'text',
            text: masterPrompt
          }
        ]
      }]
    });

    // Calculate cost from response
    if (response.usage) {
      const cost = calculateCost(response.usage, settings.model);
      totalCost += cost;
      console.log(`[COST] Request cost: $${cost.toFixed(4)} (Input: ${response.usage.input_tokens}, Output: ${response.usage.output_tokens})`);
    }

    // Step 3: Extract code
    progressCallback({ status: 'generating code', progress: 60 });

    const responseText = response.content[0].text;
    const code = extractCode(responseText);

    // Step 4: Execute code
    progressCallback({ status: 'creating document', progress: 80 });

    await executeCode(code, tempDir, fileName, progressCallback);

    const outputPath = path.join(outputDir, `${fileName}.docx`);

    // Verify file was created
    if (fs.existsSync(outputPath)) {
      console.log(`[SUCCESS] File created at: ${outputPath}`);
      const stats = fs.statSync(outputPath);
      console.log(`[SUCCESS] File size: ${stats.size} bytes`);
    } else {
      console.error(`[ERROR] File not found at expected location: ${outputPath}`);
      throw new Error(`Output file not created at: ${outputPath}`);
    }

    return {
      success: true,
      fileName: `${fileName}.docx`,
      outputPath: outputPath,
      cost: totalCost
    };

  } catch (error) {
    throw new Error(`Conversion failed: ${error.message}`);
  } finally {
    // Clean up temporary file if page selection was used
    if (tempFileCleanup) {
      tempFileCleanup();
    }
  }
}

module.exports = {
  convertFile
};

File: converter.py
"""
PDF/Image to Word converter using Claude Vision + built-in docx skill.

This module handles document conversion using:
- Claude Vision API for OCR
- Built-in docx skill for Word document generation
- Page extraction and batching for large documents
"""

import os
import base64
import logging
import tempfile
from pathlib import Path
from typing import Dict, Any, Optional, Callable
from anthropic import Anthropic
from pypdf import PdfReader, PdfWriter

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def build_prompt(settings: Dict[str, Any], file_name: str) -> str:
    """
    Build vision-optimized conversion prompt.

    Args:
        settings: User settings (font, margin, etc.)
        file_name: Output file name

    Returns:
        Complete prompt string
    """
    margin = settings.get('margin', 1.0)  # Single value now

    prompt = f"""Convert this scanned document to professional Word (.docx) format.

## Document Settings
- Font: {settings.get('font', 'Arial')} {settings.get('fontSize', 12)}pt
- Margins: {margin}" on all sides
- Output filename: {file_name}.docx

## Special Instructions"""

    if settings.get('replaceSignatures'):
        prompt += "\n- Replace signature images with text '[Signature]'"

    if settings.get('addPageMarkers'):
        prompt += "\n- Insert '[Page X of the original]' at END of complete sentences after page breaks. NEVER mid-sentence (critical for CAT tool segmentation)."

    if settings.get('customInstructions'):
        prompt += f"\n- {settings['customInstructions']}"

    prompt += """

## TEXT EXTRACTION via Vision Model
**Read the document image carefully:**
- Extract ALL text exactly as it appears (no paraphrasing)
- Preserve original language
- Handle poor scan quality, degraded text, or handwriting
- Understand document structure and layout context
- Recognize heading hierarchy, tables, lists, and formatting

## CRITICAL PRESERVATION RULES
**Reproduce EXACTLY what you see:**
1. **Exact Text**: Every word, character, number exactly as shown
2. **Exact Font Sizes**: Reproduce font sizes as-is. DO NOT interpret larger text as "headings" - just match the visual size
3. **Complete Content**: All text, tables, images, signatures, headers, footers
4. **No Skipping**: Process entire document, every page
5. **No Adding**: Do not add titles, labels, or content not visible in source
6. **No Interpretation**: Do not summarize or reword

## LAYOUT & READING ORDER
**Structured extraction:**
- **Double columns**: Extract left column first, then right column below (linear reading order for CAT tools)
- **Parallel text** (signatures, dates): Preserve spatial relationship
- **Tables**: Maintain exact structure, borders, cell content
- **Lists**: Preserve bullet/numbered formatting exactly as shown
- **Spacing**: Match paragraph spacing and line breaks

## OUTPUT QUALITY for Professional Delivery
**Create a Word document that is:**
- Clean, editable text (no text-as-images except signatures if not replaced)
- Consistent formatting throughout
- CAT-tool compatible (proper segmentation, no mid-sentence breaks)
- Minimal post-processing needed (~90% delivery-ready)
- Visually similar to original (~80-90% resemblance)

Use the docx skill to create a high-quality formatted Word document.
"""

    return prompt


def file_to_base64(file_path: str) -> str:
    """Convert file to base64 encoding"""
    with open(file_path, 'rb') as f:
        return base64.b64encode(f.read()).decode('utf-8')


def get_media_type(file_path: str) -> str:
    """Determine media type from file extension"""
    ext = Path(file_path).suffix.lower()
    media_types = {
        '.pdf': 'application/pdf',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.png': 'image/png'
    }
    return media_types.get(ext, 'application/pdf')


def extract_pages(pdf_path: str, page_range: str) -> str:
    """
    Extract specific pages from PDF

    Args:
        pdf_path: Path to PDF file
        page_range: Range string like "1-5, 7, 9-12" or empty for all

    Returns:
        Path to extracted PDF (or original if all pages)
    """
    if not page_range or page_range.strip() == '':
        return pdf_path

    reader = PdfReader(pdf_path)
    writer = PdfWriter()

    # Parse page range
    pages_to_extract = set()
    for part in page_range.split(','):
        part = part.strip()
        if '-' in part:
            start, end = part.split('-')
            pages_to_extract.update(range(int(start)-1, int(end)))  # Convert to 0-indexed
        else:
            pages_to_extract.add(int(part)-1)  # Convert to 0-indexed

    # Extract pages
    for page_num in sorted(pages_to_extract):
        if page_num < len(reader.pages):
            writer.add_page(reader.pages[page_num])

    # Save extracted PDF
    output_path = pdf_path.replace('.pdf', '_extracted.pdf')
    with open(output_path, 'wb') as f:
        writer.write(f)

    return output_path


def should_split_document(file_path: str) -> bool:
    """Determine if document needs batching"""
    file_size = os.path.getsize(file_path)

    # Split if > 25MB
    if file_size > 25 * 1024 * 1024:
        return True

    # For PDFs, check page count
    if file_path.endswith('.pdf'):
        reader = PdfReader(file_path)
        if len(reader.pages) > 90:
            return True

    return False


def create_batch_pdf(pages: list, start_idx: int) -> str:
    """Create temporary PDF from page list"""
    writer = PdfWriter()
    for page in pages:
        writer.add_page(page)

    temp_path = tempfile.mktemp(suffix=f'_batch_{start_idx}.pdf')
    with open(temp_path, 'wb') as f:
        writer.write(f)

    return temp_path


def merge_docx_files(docx_paths: list, output_name: str) -> str:
    """
    Merge multiple DOCX files preserving formatting

    Args:
        docx_paths: List of DOCX file paths
        output_name: Base name for output file

    Returns:
        Path to merged DOCX file
    """
    from docx import Document

    merged = Document(docx_paths[0])

    for docx_path in docx_paths[1:]:
        sub_doc = Document(docx_path)

        # Add page break
        merged.add_page_break()

        # Copy all elements
        for element in sub_doc.element.body:
            merged.element.body.append(element)

    output_dir = Path(docx_paths[0]).parent
    output_path = output_dir / f"{output_name}_merged.docx"
    merged.save(output_path)

    # Clean up batch files
    for path in docx_paths:
        try:
            os.unlink(path)
        except:
            pass

    return str(output_path)


def convert_document_batched(
    file_path: str,
    settings: Dict[str, Any],
    api_key: str,
    progress_callback: Optional[Callable] = None,
    client: Optional[Anthropic] = None
) -> Dict[str, Any]:
    """
    Convert large PDFs in batches (15 pages each)

    Args:
        file_path: Path to PDF file
        settings: User settings
        api_key: Anthropic API key
        progress_callback: Optional callback
        client: Optional pre-configured client

    Returns:
        Same as convert_document_simple
    """
    try:
        if progress_callback:
            progress_callback({'status': 'preparing batches', 'progress': 5})

        reader = PdfReader(file_path)
        total_pages = len(reader.pages)
        batch_size = 15
        batches = []

        # Create batches
        for i in range(0, total_pages, batch_size):
            batch_pages = reader.pages[i:i + batch_size]
            batches.append(batch_pages)

        logger.info(f"Splitting {total_pages} pages into {len(batches)} batches")

        converted_paths = []
        total_cost = 0.0

        # Process each batch
        for idx, batch_pages in enumerate(batches):
            batch_progress = 10 + (80 * idx / len(batches))

            if progress_callback:
                progress_callback({
                    'status': f'converting batch {idx+1}/{len(batches)}',
                    'progress': batch_progress
                })

            # Create temp PDF for batch
            batch_pdf_path = create_batch_pdf(batch_pages, idx)

            # Convert batch
            result = convert_document_simple(
                batch_pdf_path,
                settings,
                api_key,
                None,  # No progress callback for batches
                client
            )

            # Cleanup batch PDF
            try:
                os.unlink(batch_pdf_path)
            except:
                pass

            if not result['success']:
                # Cleanup on failure
                for path in converted_paths:
                    try:
                        os.unlink(path)
                    except:
                        pass
                return result

            converted_paths.append(result['output_path'])
            total_cost += result.get('cost', 0)

        # Merge all batches
        if progress_callback:
            progress_callback({'status': 'merging batches', 'progress': 90})

        file_name = Path(file_path).stem
        merged_path = merge_docx_files(converted_paths, file_name)

        if progress_callback:
            progress_callback({'status': 'complete', 'progress': 100})

        return {
            'success': True,
            'output_path': merged_path,
            'cost': total_cost
        }

    except Exception as e:
        logger.error(f"Batched conversion failed: {e}")
        return {
            'success': False,
            'error': str(e)
        }


def convert_document(
    file_path: str,
    settings: Dict[str, Any],
    api_key: str,
    page_range: str = '',
    skill_id: Optional[str] = None,
    progress_callback: Optional[Callable] = None,
    client: Optional[Anthropic] = None
) -> Dict[str, Any]:
    """
    Convert document to Word format using Claude Vision + docx skill.

    Args:
        file_path: Path to PDF or image file
        settings: User settings (font, margin, model, etc.)
        api_key: Anthropic API key
        page_range: Page range for PDFs (e.g., "1-5, 7")
        skill_id: Ignored (kept for compatibility)
        progress_callback: Optional callback(dict) for progress updates
        client: Optional pre-configured client (for testing)

    Returns:
        Dict with:
            - success: bool
            - output_path: str (path to generated .docx)
            - cost: float (in USD)
            - error: str (if failed)
    """
    try:
        if progress_callback:
            progress_callback({'status': 'preparing', 'progress': 10})

        # Extract pages if range specified
        if page_range and file_path.endswith('.pdf'):
            file_path = extract_pages(file_path, page_range)

        # Check if batching needed
        if should_split_document(file_path):
            logger.info("Document requires batching")
            return convert_document_batched(
                file_path, settings, api_key, progress_callback, client
            )
        else:
            # Simple conversion for small documents
            return convert_document_simple(
                file_path, settings, api_key, progress_callback, client
            )

    except Exception as e:
        logger.error(f"Conversion failed: {e}")
        return {
            'success': False,
            'error': str(e)
        }


def convert_document_simple(
    file_path: str,
    settings: Dict[str, Any],
    api_key: str,
    progress_callback: Optional[Callable] = None,
    client: Optional[Anthropic] = None
) -> Dict[str, Any]:
    """Simple conversion for small documents using Vision + docx skill"""

    if progress_callback:
        progress_callback({'status': 'preparing', 'progress': 20})

    # Initialize client
    if client is None:
        client = Anthropic(api_key=api_key)

    # Prepare file
    file_base64 = file_to_base64(file_path)
    media_type = get_media_type(file_path)
    file_name = Path(file_path).stem

    # Build prompt - use custom if available
    from config_manager import ConfigManager
    config_mgr = ConfigManager()
    custom_prompt = config_mgr.get_custom_prompt()

    if custom_prompt:
        # Use custom prompt, but inject filename
        prompt = custom_prompt.replace('{file_name}', file_name)
    else:
        # Use default prompt builder
        prompt = build_prompt(settings, file_name)

    if progress_callback:
        progress_callback({'status': 'analyzing', 'progress': 40})

    # API call
    logger.info(f"Calling Anthropic API with model {settings.get('model')}")

    response = client.beta.messages.create(
        model=settings.get('model', 'claude-sonnet-4-5-20250929'),
        max_tokens=16000,
        betas=['code-execution-2025-08-25', 'skills-2025-10-02', 'files-api-2025-04-14'],
        container={
            'skills': [{
                'type': 'anthropic',
                'skill_id': 'docx',
                'version': 'latest'
            }]
        },
        messages=[{
            'role': 'user',
            'content': [
                {
                    'type': 'document',
                    'source': {
                        'type': 'base64',
                        'media_type': media_type,
                        'data': file_base64
                    }
                },
                {
                    'type': 'text',
                    'text': prompt
                }
            ]
        }],
        tools=[{
            'type': 'code_execution_20250825',
            'name': 'code_execution'
        }]
    )

    if progress_callback:
        progress_callback({'status': 'generating', 'progress': 70})

    # Extract file IDs
    file_ids = []
    for item in response.content:
        if item.type == 'bash_code_execution_tool_result':
            content_item = item.content
            if content_item.type == 'bash_code_execution_result':
                for file in content_item.content:
                    if hasattr(file, 'file_id'):
                        file_ids.append(file.file_id)

    if not file_ids:
        return {
            'success': False,
            'error': 'No file generated by skill'
        }

    # Download file
    file_id = file_ids[0]
    file_metadata = client.beta.files.retrieve_metadata(
        file_id=file_id,
        betas=["files-api-2025-04-14"]
    )
    file_content = client.beta.files.download(
        file_id=file_id,
        betas=["files-api-2025-04-14"]
    )

    # Save file
    output_dir = Path(file_path).parent
    output_path = output_dir / file_metadata.filename
    file_content.write_to_file(output_path)
    logger.info(f"Downloaded: {file_metadata.filename}")

    if progress_callback:
        progress_callback({'status': 'complete', 'progress': 100})

    # Calculate cost
    from cost_calculator import calculate_actual_cost
    cost = calculate_actual_cost(
        {
            'input_tokens': response.usage.input_tokens,
            'output_tokens': response.usage.output_tokens
        },
        settings.get('model')
    )

    return {
        'success': True,
        'output_path': str(output_path),
        'cost': cost
    }
